pegasus/helpers/form_helpers.rb

```diff
@@ -54,16 +54,12 @@ def insert_or_upsert_form(kind, data, options={})
   form_class = Object.const_get(kind)
   row[:source_id] = form_class.get_source_id(data) if form_class.respond_to? :get_source_id
 
-  # For HOC signups, a duplicate (matching email and name) entry is assumed to be an update rather
-  # than an insert.
-  if kind.start_with? 'HocSignup'
-    row_for_update = row.dup
-    [:created_at, :created_ip, :secret].each {|key| row_for_update.delete key}
-    num_rows_updated = DB[:forms].
-      where(kind: kind, email: row_for_update[:email], name: row_for_update[:name]).
-      update(row_for_update)
-    return row_for_update if num_rows_updated > 0
+  # Depending on the form kind, new data, and existing data, perform an update.
+  existing_form_secret = form_class.update_on_upsert row
+  if existing_form_secret
+    return update_form(row[:kind], existing_form_secret, row)
   end
+
```

Comment:
Fortunately (unfortunately?) not, since we only update the DB with the contents of `form`, not `data`. :(